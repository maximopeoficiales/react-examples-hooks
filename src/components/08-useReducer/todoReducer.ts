export enum Action {
    ADD = "add",
    DELETE = "delete",
    TOGGLE = "toggle",
    TOGGLE_OLD = "toggle_old",
    EMPTY = "",
}
export type Todo =
    {
        id: string;
        desc: string;
        done: boolean;
    }

type ReducerTodo = { type: Action, payload: Todo };

export const todoReducer = (state: Todo[], action: ReducerTodo): Todo[] => {
    switch (action.type) {
        case Action.ADD:
            return [...state, action.payload];

        case Action.DELETE:
            return state.filter(todo => todo.id !== action.payload.id);

        case Action.TOGGLE:
            return state.map((todo) => (todo.id === action.payload.id) ? { ...todo, done: !todo.done } : todo);

        case Action.TOGGLE_OLD:
            return state.map((todo) => {
                if (todo.id === action.payload.id) {
                    return { ...todo, done: !todo.done };
                } else {

                    return todo;
                }
            });
        default:
            return state;
    }
}