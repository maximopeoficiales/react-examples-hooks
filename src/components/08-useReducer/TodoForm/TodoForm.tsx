import { useForm } from "../../../hooks/useForm";
import { Todo } from "../todoReducer";

interface MyProps {
  addTodo(todo: Todo): void;
}
const defaultProps: MyProps = {
  addTodo: () => {},
};
const TodoForm = (props: MyProps) => {
  props = { ...defaultProps, ...props };
  const {
    formValues: { description },
    handlerChange,
  } = useForm({ description: "" });

  const handlerSubmit = (e: React.ChangeEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (description.trim().length <= 1) return;
    const newTodo: Todo = {
      id: new Date().getTime().toString(),
      desc: description,
      done: false,
    };
    props.addTodo(newTodo);
  };

  return (
    <form action="" onSubmit={handlerSubmit}>
      <div className="form-group">
        <input
          type="text"
          name="description"
          className="form-control"
          placeholder="Insert todo"
          value={description}
          onChange={handlerChange}
        />
      </div>
      <button className="btn btn-warning d-block btn-lg my-2 w-100">Add</button>
    </form>
  );
};

export default TodoForm;
