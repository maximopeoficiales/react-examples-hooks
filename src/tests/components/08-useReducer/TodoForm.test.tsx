import { shallow } from "enzyme";
import React, { ChangeEvent } from "react";
import TodoForm from "../../../components/08-useReducer/TodoForm/TodoForm";
import { Todo } from "../../../components/08-useReducer/todoReducer";

describe("Pruebas en TodoForm", () => {
  const addTodo = jest.fn();

  const wrapper = shallow(<TodoForm addTodo={addTodo} />);

  test("debe mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("No debe de llamar el addTodo", () => {
    const formSubmit: any = wrapper.find("form").prop("onSubmit");

    expect(addTodo).toHaveBeenCalledTimes(0);
  });
  test("debe de llamar la funcion addTodo", () => {
    const value = "nueva descripcion";
    const event: any = {
      target: {
        name: "description",
        value,
      },
    };
    wrapper.find("input").simulate("change", event);

    // console.log(wrapper.find("input").html());
    // se ejecuta el submit
    const formSubmit: any = wrapper.find("form").prop("onSubmit");
    formSubmit({ preventDefault() {} });

    expect(addTodo).toHaveBeenCalledTimes(1);
    // se espera que se llame esta funcion con cualquier objecto como parametro
    expect(addTodo).toHaveBeenCalledWith(expect.any(Object));
    // se espera que la funciona reciba como parametro un todo
    expect(addTodo).toHaveBeenCalledWith({
      id: expect.any(String),
      desc: value,
      done: false,
    });

    // se espera que se hallan reseteado las propiedades
    expect(wrapper.find("input").prop("value")).toHaveLength(0);
  });
});
