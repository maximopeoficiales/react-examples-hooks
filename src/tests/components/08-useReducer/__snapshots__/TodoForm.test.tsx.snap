// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoForm debe mostrarse correctamente 1`] = `
<form
  action=""
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <input
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Insert todo"
      type="text"
      value=""
    />
  </div>
  <button
    className="btn btn-warning d-block btn-lg my-2 w-100"
  >
    Add
  </button>
</form>
`;
